base/shell/explorer/desktop.cpp

```diff
@@ -22,107 +22,184 @@
-        CComPtr<IShellDesktopTray> pSdt;
-        HANDLE hDesktop;
-        HRESULT hRet;
-
-        OleInitialize(NULL);
-
-        hRet = m_Tray->QueryInterface(IID_PPV_ARG(IShellDesktopTray, &pSdt));
-        if (!SUCCEEDED(hRet))
-            return 1;
-
-        hDesktop = _SHCreateDesktop(pSdt);
-        if (hDesktop == NULL)
-            return 1;
+    DWORD DesktopThreadProc();
+    static DWORD WINAPI s_DesktopThreadProc(LPVOID lpParameter);
+    
+public:
+    
+    CDesktopThread();
+    virtual ~CDesktopThread();
 
-        if (!SetEvent(m_hEvent))
-        {
-            /* Failed to notify that we initialized successfully, kill ourselves
-            to make the main thread wake up! */
-            return 1;
-        }
+    HRESULT Initialize(ITrayWindow* pTray);
+    void Destroy();
+};
 
-        _SHDesktopMessageLoop(hDesktop);
+/*******************************************************************/
 
-        /* FIXME: Properly rundown the main thread! */
-        ExitProcess(0);
+CDesktopThread::CDesktopThread() :
+    m_Tray(NULL),
+    m_hEvent(NULL),
+    m_hThread(NULL)
+{
+}
 
-        return 0;
+CDesktopThread::~CDesktopThread()
+{
+    if (m_hEvent || m_hThread)
+    {
+        Destroy();
     }
+}
 
-    static DWORD CALLBACK s_DesktopThreadProc(IN OUT LPVOID lpParameter)
+HRESULT CDesktopThread::Initialize(ITrayWindow* pTray)
+{
+    HANDLE hEvent;
+    HANDLE hThread;
+    HANDLE Handles[2];
+    
+    if (!pTray)
     {
-        return reinterpret_cast<CDesktopThread*>(lpParameter)->DesktopThreadProc();
+        return E_FAIL;
     }
-
-public:
-    CDesktopThread() :
-        m_hEvent(NULL),
-        m_Tray(NULL)
+    
+    if (m_hEvent || m_hThread)
     {
+        return E_FAIL;
     }
+    
+    hEvent = CreateEventW(NULL, FALSE, FALSE, NULL);
 
-    HRESULT Initialize(IN OUT ITrayWindow *pTray)
+    if (!hEvent)
     {
-        HANDLE hThread;
-        HANDLE Handles[2];
+        return E_FAIL;
+    }
+
+    hThread = CreateThread(NULL, 0, s_DesktopThreadProc, (LPVOID)this, CREATE_SUSPENDED, NULL);
 
-        m_Tray = pTray;
+    if (!hThread)
+    {   
+        CloseHandle(hEvent);
+        
+        return E_FAIL;
+    }
+    
+    m_Tray = pTray;
+    m_hEvent = hEvent;
+    m_hThread = hThread;
+    
+    ResumeThread(hThread);
+
+    Handles[0] = m_hThread;
+    Handles[1] = m_hEvent;
 
-        m_hEvent = CreateEventW(NULL, FALSE, FALSE, NULL);
-        if (!m_hEvent)
-            return E_FAIL;
+    for (;;)
+    {
+        DWORD WaitResult = MsgWaitForMultipleObjects(_countof(Handles), Handles, FALSE, INFINITE, QS_ALLEVENTS);
 
-        hThread = CreateThread(NULL, 0, s_DesktopThreadProc, (PVOID)this, 0, NULL);
-        if (!hThread)
+        if (WaitResult == WAIT_OBJECT_0 + _countof(Handles))
         {
-            CloseHandle(m_hEvent);
-            return E_FAIL;
+            TrayProcessMessages(m_Tray);
         }
+        
+        else if (WaitResult != WAIT_FAILED && WaitResult != WAIT_OBJECT_0)
```

Comment:
I know you didn't change this, but maybe you understand: why is this loop not exiting in the `WAIT_OBJECT_0` case? If the thread encounters a failure, it seems to exit without setting the event, so it seems like we'd loop forever here in that case.

Reply:
I think I need to rewrite the class because an old one used ExitProcess() hack.